### ISSUE TITLE
Sharing: Easy bookmarking of shared albums

### ISSUE THREAD

**Issue by inthreedee** [photoprism/photoprism#3133]
#### 1. What is not working as documented?

- Sharing an album generates a URL such as https://photos.example.com/s/shared-key/album-name.
- I shared such a link with family members to view a shared album. When they click the link and view the album, the link changes to something else, such as https://photos.example.com/library/albums/some-other-key/album-name, which appears to be the internal URL for that album.
- They then bookmark that URL to return later. Obviously, it doesn't work in a new session since only the shared URL works.
- Explaining to them how to manually create a bookmark containing the correct URL so it works every time is a bit too much for me to expect them to be able to handle.

#### 2. How can we reproduce it?

1. Share an album and copy its shared URL
2. Visit the shared URL
3. Shared URL changes to the album's internal URL after visiting
4. Bookmark the page, start a new browser session, and try returning to the bookmark.

#### 3. What behavior do you expect?

When they visit the shared URL, the URL should not redirect to the internal, permanent URL of the album. That way, they can easily bookmark the shared URL and return to it in the future.

#### 4. What could be the cause of your problem?

Shared URLs redirecting to internal URLs?

#### 5. Can you provide us with example files for testing, error logs, or screenshots?

Shared URL format:
https://photos.example.com/s/shared-key/album-name
Changes to this format after the page loads:
https://photos.example.com/library/albums/some-other-key/album-name

#### 6. Which software versions do you use?

(a) PhotoPrism Architecture & Build Number: ARM64 Build 221118-e58fee0fb

(b) Database Type & Version: MariaDB 10.9

(c) Operating System Types & Versions: Linux

(d) Browser Types & Versions: Firefox

(e) Ad Blockers, Browser Plugins, and/or Firewall Software? ?

#### 7. On what kind of device is PhotoPrism installed?

This is especially important if you are reporting a performance, import, or indexing issue. You can skip this if you're reporting a problem you found in our public demo, or if it's a completely unrelated issue, such as incorrect page layout.

(a) Device / Processor Type: Raspberry Pi 4

(b) Physical Memory & Swap Space in GB: 8GB RAM, 4GB Swap

(c) Storage Type: SSD

#### 8. Do you use a Reverse Proxy, Firewall, VPN, or CDN?

Reverse proxy using https://hub.docker.com/r/jc21/nginx-proxy-manager/ in its own docker container

**Comment by lastzero** [comment]
I get your use case, but it's actually a hidden feature in that it's possible to limit the number of times a link can be used to prevent people from sharing the links with the entire internet. Unfortunately, we didn't have any time to add this to the sharing form nor document it. Also I believe there's already a feature request for reusable bookmarks. On my phone right now, so can't search.

**Comment by inthreedee** [comment]
> it's actually a hidden feature in that it's possible to limit the number of times a link can be used to prevent people from sharing the links with the entire internet

If someone we share it with wants to share the URL with the entire internet, they can just copy/paste the share link as long as it's valid, so I don't see how this actually prevents any unwanted behavior. The real solution to that problem already exists; we have the ability to set expiration dates and change the sharing key whenever we want.

Creating this weird redirect scenario prevents legitimate users from bookmarking shared albums. It seems like this causes more harm than good and acts to defeat the primary purpose of the shared albums feature. As it is now, they're basically single-use shares unless someone is tech-savvy enough to recognize the redirect scenario and manually save the share link to a bookmark. I'd wager that's a very small percentage of the recipients of a shared album link.


I did a search before posting and found this request, which didn't seem to be about shared album URLs:
https://github.com/photoprism/photoprism/issues/831, and which was closed in favor of https://github.com/photoprism/photoprism/issues/1307.

**Comment by inthreedee** [comment]
For extra context, I very nearly just wiped all my shared albums from Google Photos in favor of my Photoprism instance, a process I've been working towards for some time now. I didn't expect shared family albums to break in this way and it's a huge wrinkle in my plans to ditch Google Photos. I consider my family members being able to bookmark an album that I share with them a basic feature of any photo sharing tool.

There is a certain amount of trust that's expected when we share albums with someone, so therefore I don't think it's necessary or reasonable for a photo sharing tool to restrict legitimate sharing in this way.

Google Photos shared albums have URLs in this format, which do not redirect after visiting:
https://photos.google.com/share/sharing-key

And the internal link looks like so:
https://photos.google.com/album/internal-key

**Comment by lastzero** [comment]
Yes, right now that's possible as there simply is no input field to set a limit. Since the album then is bound to your secret session id and not the original link anymore, it is in fact not possible to share the link any further. Of course you could download all the pictures and share those on a different site.

**Comment by inthreedee** [comment]
I understand what you're saying about the internal URL not being sharable, but the share link is still always sharable. So this hidden feature you're talking about fails to prevent the behavior you're saying it's designed to prevent because, again, the original share URL is still valid and sharable.

I'm arguing that this "hidden feature" is worthless and only gets in the way of legitimate sharing.

If we think through the scenario you're trying to prevent:
- I share an album with Bob.
- Bob visits the album and decides to share the album on facebook against my wishes.
- Bob copies the url, which is invalid for any sessions outside his own, and shares it
- Someone says "hey, that link doesn't work!"
- Bob realizes oh, there must be something special about the share link and posts the original share link to facebook instead.
- Done, everyone has access anyway.

I'd argue that it takes very little tech savvy for Bob to make that leap and, therefore, any normal person would follow those same steps if their goal was to share the album on the wider internet.

The primary use-case for sharing albums with family members is for them to be able to return to those shared albums later and continue viewing them. I'm sure single-use shares have some value, but I imagine most people sharing albums with family want their family to be able to return to those albums, and bookmarks are the way that normal people achieve this. So actively preventing bookmarking breaks the shared albums feature in this case.

With the current behavior, I'll have to abandon my plans to use Photoprism to share albums with the family and retain my Google Photos paid plan instead. It's actually that big of a problem and I imagine I can't be the only person who will run into this problem with shared albums.

**Comment by lastzero** [comment]
It may be worthless to you, which is why I said in the beginning that I get your point. Given that we don't even have time to add a simple input field, there is now way we can implement a completely different functionality as a spontaneous reaction to your "bug" report.

**Comment by inthreedee** [comment]
> as a spontaneous reaction to your "bug" report

Okay, let's take a breath. I'm sorry if my frustration is coming through, I'm really not trying to instigate anything here. I'm trying to be helpful.

I'm finding it very difficult to understand the decision to make this kind of tradeoff and so I'm trying to, as politely as possible, encourage a constructive conversation about the value and tradeoffs of this hidden feature. I just have a hard time wrapping my head around the logic here and I can't believe I'm in the minority in expecting shared album URLs to function the way I'm describing, similarly to other photo tools.

But it sounds like the decision is made, so alright. Feel free to close the issue as intended behavior.

**Comment by lastzero** [comment]
As I said, we completely get your point and see the value. The only problem is that I'm a single developer and there are over 25,000 users. This is extremely frustrating for us too, as we could do much better with proper funding. This is why we focus on funding now before implementing feature after feature in an unsustainable death march.

**Comment by pgalbavy** [comment]
Just to add my view, I have to agree with @inthreedee - I was just surprised by this after taking photos for my company Xmas party and sharing album links. I did not expect this. I agree however that it's not a bug, it's a feature - just not very clear.

Can I suggest as another alternative an option during the share process to also make an album "public" (which is probably part of a different request)? Later the owner can make an album private again, if needed.

**Comment by lastzero** [comment]
@pgalbavy Thanks for sharing your thoughts on this! I agree that this should be improved and that it should be possible to create share links that don't require any kind of session, so basically like this feature request (which is already waiting way too long to get implemented):
- https://github.com/photoprism/photoprism/issues/669

**Comment by seanwalter** [comment]
I agree it would be good to have sharing URLs bookmark-able and re-share-able from the browser.

I have had a few situations with my wife or a friend where I’ve sent them a, “Hey, test this shared link and if it works for you send it out,” and what they send along is the URL from the browser location bar, not the original URL.

Going forward I can try to remember that this feature exists and needs to be managed, it’s just not intuitive.

**Comment by lastzero** [comment]
The share URLs are bookmarkable, but the URL you are redirected to is not a public URL. This is because the implementation supports a view counter / limit that should only be increased / checked when you use a link for the first time, but not every time you return to the overview.

**Comment by seanwalter** [comment]
Thank you for clarifying. It’s the redirect to the non-public URL that leads to the counterintuitive user experience.

Most people who visit a web page that didn’t require a password assume they can bookmark it while viewing it, and then go back to it later. They generally also assume they can copy the URL in their location bar or use the browser’s “share” feature to forward the page to others.

I appreciate all you are trying to manage here and am a big fan of PhotoPrism. I am just weighing in with others that it would be great for the shared URL experience to be more intuitive for people.

Thank you!

**Comment by lastzero** [comment]
Totally understand, thus we need the public image wall feature which will be a static URL unlike the current system which is more like in invite code, so you can limit use without blocking access for existing users. Note that these can also bookmark the URL as long as they keep using the same browser. Sending the link to other users won't work though and that's the point. Also the browser session will eventually expire, so you need to use the original invite link again to get access.
